### ISSUE TITLE
export in library works only for last entry module

### ISSUE THREAD

**Issue by vankop** [webpack/webpack#15936]
# Bug report

export in library works only for last entry module

**What is the current behavior?**

in webpack config like:
```js
module.exports = {
  entry: ["a", "b", "c"],
  output: {
    library: { type: "global" }
  }
};
```

only exports from `c` will be used

**If the current behavior is a bug, please provide the steps to reproduce.**

https://github.com/TGRHavoc/webpack-mvp

**What is the expected behavior?**
should use all exports

**Other relevant information:**
webpack version:  latest
Node.js version: 
Operating System: 
Additional tools:

**Comment by alexander-akait** [comment]
bump

**Comment by alan-agius4** [comment]
Is there any workaround for this? Other than creating a barrel?

**Comment by alexander-akait** [comment]
@alan-agius4 currently, no workaround (only multi compiler builds), but feel free to fix send a PR with fix

**Comment by TGRHavoc** [comment]
> @alan-agius4 currently, no workaround (only multi compiler builds), but feel free to fix send a PR with fix

Does anyone know where one would get started at debugging/fixing this issue?

**Comment by Judahmeek** [comment]
@TGRHavoc I'd start by checking what the default for the `output.library.name` configuration is.

A client of mine used [some custom output.library.name logic back in the day](https://rossta.net/blog/from-sprockets-to-webpack.html#exporting-from-webpack) to support multiple library sub-modules.

**Comment by Elessar1802** [comment]
I was trying to replicate this issue and noticed as per the documentation, we are multi-part entries are used in this case - https://github.com/webpack/webpack/tree/main/examples/multi-part-library

So this seems like a intentional design choice. Don't you think?

Workarounds
1. Barrel file
2. Add the following to `index.html`
```
        <script src="dist/index.js"></script>
        <script src="dist/scriptB.js"></script>
        <script src="dist/scriptC.js"></script>
```

@TGRHavoc

**Comment by TGRHavoc** [comment]
> I was trying to replicate this issue and noticed as per the documentation, we are multi-part entries are used in this case - https://github.com/webpack/webpack/tree/main/examples/multi-part-library
> 
> So this seems like a intentional design choice. Don't you think?
> 
> Workarounds
> 1. Barrel file
> 2. Add the following to `index.html`
> ```
>         <script src="dist/index.js"></script>
>         <script src="dist/scriptB.js"></script>
>         <script src="dist/scriptC.js"></script>
> ```
> 
> @TGRHavoc 
> 

Sorry, my memory is a little fuzzy on this issue (I'm trying to use the discussion thread to re-jog my memory https://github.com/orgs/webpack/discussions/15929).
I think the issue was around exposing the exported functions to the HTML from multiple files. As far as I could find out, exporting everything to a global library _should_ have done exactly this but it wouldn't. 
Because I was trying to modernise a very large legacy system, adding additional scripts to every page wasn't feasible and neither was barrel files. In the end, I didn't touch the older stuff and only modernised newer functionality because of this friction with webpack (I originally decided to use webpack because I thought it had this functionality and would allow me to move forward whilst keeping the system functioning). 

Basically, if webpack exposed all exported functions for a named global library, I would expect it to also expose all exported functions for a non-named library. The fact that _only_ the last entry gets exposed does feel like a bug and not intented (maybe it should be all or none?)

**Comment by webdevelopersrinu** [comment]
Hi I'd like to work on this issue.

I've traced the root cause: in `AbstractLibraryPlugin.js`, `finishEntryModule` is only called on the last entry dependency (`deps[deps.length - 1]`), so exports from earlier entries in the array are never flagged as used and get dropped from the library output.

My plan is to call `finishEntryModule` for all entry dependencies and adjust the startup runtime so the library exposes the merged exports of all entry modules, plus add test cases under `test/configCases/library`. Please let me know if this direction sounds right — happy to adjust based on feedback.

**Comment by webdevelopersrinu** [comment]
I've implemented the approach described in my earlier comment — fix in the library plugin layer, test case included, zero regressions in the library test suite. The PR (#21662) was auto-closed by the quality bot, which appears to be flagging my contributor history rather than the change itself (CLA signed, changeset included). Would a maintainer be willing to take a look and reopen if the approach looks right?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

I've read through the report. Before changing anything I'd like to reproduce it against current `main` first — my reasoning is that a fix anchored to the actual observed behavior is much easier to review than one inferred from the report alone, especially if the surrounding code has moved since this was filed.

Plan: reproduce and confirm the root cause, then make the smallest change that addresses it (plus a regression test if the repo has coverage for this area). If the root cause turns out to live somewhere other than where the report points, I'll report back here before forcing a change.

Anything already known about this that I should read first?

(claiming via @LeonxLJX)
